### PR TITLE
[Python] Add support for `datetime64` columns with a time-unit that is not `ns`

### LINKED ISSUE #8588: Pandas DataFrame returns incorrect datetime values (date = epoch zero point / 1970-01-01)
### What happens?

I use duckdb to work with pandas.  I find duckdb return wrong datetime after execute sql on the DataFrame. 
The following pandas DataFrame with name 'stdd' has a column 'datetime'
```
     datetime  voli
0 2005-01-04   NaN
1 2005-01-05   0.0
2 2005-01-06   0.0
3 2005-01-07   0.0
4 2005-01-10   0.0
```
the result of `con.execute("select * from stdd").df()` is:
```
                     datetime  voli
0 1970-01-01 00:00:01.104796   NaN
1 1970-01-01 00:00:01.104883   0.0
2 1970-01-01 00:00:01.104969   0.0
3 1970-01-01 00:00:01.105056   0.0
4 1970-01-01 00:00:01.105315   0.0
```
I have tried the latest master build version: 0.8.2-dev2850

### To Reproduce

Need the pickle file to reproduce : [stdd.tar.gz](https://github.com/duckdb/duckdb/files/12356094/stdd.tar.gz)
I save the pandas DataFrame 'stdd' in pickle file, and tar it for uploading policy.

1. dowdload the gz file [stdd.tar.gz](https://github.com/duckdb/duckdb/files/12356094/stdd.tar.gz)，and tar it to pickle file
```
tar xzvf stdd.tar.gz
```
2. run the following python script to reproduce:
```
import duckdb
import pickle
con=duckdb.connect()
print("duckdb version: ",duckdb.__version__)

with open("stdd.pk","rb") as f:
    stdd=pickle.load(f)

stdd_df=con.execute("select * from stdd").df()
print("DataFrame stdd:\n",stdd)
print("DataFrame stdd from duckdb:\n",stdd_df)
```
3.  output:
```
duckdb version:  0.8.2-dev2850
DataFrame stdd:
   Paste
0  Select All
1
2  How-to disable mouse
3 2005-01-07   0.0
4 2005-01-10   0.0
DataFrame stdd from duckdb:
                     datetime  voli
0 1970-01-01 00:00:01.104796   NaN
1 1970-01-01 00:00:01.104883   0.0
2 1970-01-01 00:00:01.104969   0.0
3 1970-01-01 00:00:01.105056   0.0
4 1970-01-01 00:00:01.105315   0.0
```

### OS:

debian 11

### DuckDB Version:

0.8.2-dev2850

### DuckDB Client:

Python

### Full Name:

yang

### Affiliation:

qc

### Have you tried this on the latest `master` branch?

I have tested with a master build

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] Yes, I have

### PR DESCRIPTION
**Pull request by Tishj**
This PR fixes #8588 

In Pandas < 2.0.0, Pandas would never produce a `datetime64` that is not represented in nanoseconds.
Since 2.0.0, they have added support for other units.

This PR adds support for scanning and producing columns of `datetime64` with any unit.
Pandas timestamp units directly map to our timestamp types, with the exception of `TIMESTAMP_TZ`.

`TIMESTAMP_TZ` is represented in US (microseconds), so whenever we encounter a `datetime64` column that has `tz` info, we resort back to the old approach of converting the value to microseconds.

### CODE REVIEW

File: tools/pythonpkg/src/include/duckdb_python/numpy/numpy_type.hpp [hunk 1]
```diff
@@ -12,34 +12,43 @@
...
+	DATETIME_MS, //! datetime64[ms], <M8[ms]
+	DATETIME_NS, //! datetime64[ns], <M8[ns]
+	DATETIME_US, //! datetime64[us], <M8[us]
+	TIMEDELTA,   //! timedelta64[D], timedelta64
 
 	//! ------------------------------------------------------------
 	//! Extension Types
 	//! ------------------------------------------------------------
-	CATEGORY,    //! category
-	DATETIME_TZ, //! datetime64[ns, TZ]
```
Comment (hawkfish) [inline tools/pythonpkg/src/include/duckdb_python/numpy/numpy_type.hpp:41]: Why was this removed?

File: tools/pythonpkg/src/numpy/type.cpp [hunk 2]
```diff
@@ -6,72 +6,114 @@
 
 namespace duckdb {
 
-static bool IsDateTime(const string &col_type_str) {
-	if (StringUtil::StartsWith(col_type_str, "datetime64[ns")) {
+static bool IsDateTime(NumpyNullableType type) {
+	if (type == NumpyNullableType::DATETIME_NS) {
```
Comment (hawkfish) [inline tools/pythonpkg/src/numpy/type.cpp:10]: switch statement?

**Review by hawkfish — APPROVED** [review]

File: tools/pythonpkg/src/numpy/numpy_scan.cpp [hunk 3]
```diff
@@ -234,20 +234,61 @@ void NumpyScan::Scan(PandasColumnBindData &bind_data, idx_t count, idx_t offset,
...
-	case NumpyNullableType::DATETIME_TZ: {
+	case NumpyNullableType::DATETIME_NS:
+	case NumpyNullableType::DATETIME_MS:
+	case NumpyNullableType::DATETIME_US:
+	case NumpyNullableType::DATETIME_S: {
 		auto src_ptr = reinterpret_cast<const int64_t *>(array.data());
 		auto tgt_ptr = FlatVector::GetData<timestamp_t>(out);
 		auto &mask = FlatVector::Validity(out);
 
+		// FIXME: 'has_timezone' is unused still, we should probably extract the timezone and convert the value back to
```
[the remark this answers is not in the document]
Comment (Tishj) [reply tools/pythonpkg/src/numpy/numpy_scan.cpp:245]: Yea I think this is old, the timezone conversion is done at a later time I think

File: tools/pythonpkg/src/include/duckdb_python/numpy/numpy_type.hpp (hunk 1 above)
Comment (Tishj) [reply tools/pythonpkg/src/include/duckdb_python/numpy/numpy_type.hpp:41]: It's removed in favor of `has_timezone`, previously we only ever received a single time unit from pandas, ns, now that we can receive multiple we have to remember which unit this is, together with the fact that it has timezone information

**Review by hawkfish — APPROVED** [review]

**Comment by hannes** [comment]
Thanks

**Pull request merged**